### PR TITLE
Track dropped spans and logs due to full buffer

### LINKED ISSUE #2273: Internal logging in BatchProcessor when buffer full
Opening a separate issue to specifically tracking internal logging when BatchProcessor gets full.
https://github.com/open-telemetry/opentelemetry-rust/issues/2175#issuecomment-2394816345

1. We certainly don't want to log each time the buffer is full
2. A good balance might be to keep a count of items-dropped-due-to-buffer-full.
3. This count can be exposed as a Metric itself.
4. And/Or this count can be emitted as a internal log upon shutdown/drop with severity WARN if non-zero items were dropped.

2,4 is what OTel .NET has implemented: See https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/BatchExportProcessor.cs#L102 and https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/BatchExportProcessor.cs#L207

### PR DESCRIPTION
**Pull request by scottgerring**
Fixes #2273

Design discussion issue (if applicable) #

## Changes

As suggested in #2273, this change adds a counter and a metric for dropped spans and logs; the counter is used to log total drop count at exporter shutdown, the metric is used to export this information.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:315]: I'd suggest holding off from exposing these via OTel metrics.
1. Fetching meter from global only works if user has already set a global meter provider. There is no document guiding them to setup Metrics signal before logs/traces, so by default, this is not guaranteed to work.
2. We cannot guarantee the name of the metric nor the name of the attributes. OTel spec/convention is planning to add one for the same, and we need to wait until it lands (We can offer OTel Rust as one of the language to prototype this to support spec/semantic convention)

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:182]: for the very first time, lets also emit an internal log informing the user about data drop.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:204]:
```suggestion
                name: "BatchLogProcessor.LogsDropped",
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:205]:
```suggestion
                dropped_logs_count= dropped_logs,
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:206]:
```suggestion
                message = "Logs were dropped due to full buffer. The count represents the total count of lost logs in the lifetime of the BatchLogProcessor. Consider increasing the queue size and/or decrease delay between intervals."
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:207]: would be nice to also include an attribute for the max queue size the batchprocessor was operating with.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:206]: I am unsure if it is correct to state that logs were dropped due to Buffer full. It looks like there could be other reasons too?

**Review by cijothomas — CHANGES_REQUESTED** [review]
Thanks for working on this! Left my comments in the PR.
OTel internal metrics is something we can add, but not right away:
https://github.com/open-telemetry/opentelemetry-rust/pull/2357/files#r1860843464

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [reply opentelemetry-sdk/src/logs/log_processor.rs:182]: Added this to each processor:

https://github.com/open-telemetry/opentelemetry-rust/blob/b626d7f5b00c336b0d1c263df9e43521269d8bee/opentelemetry-sdk/src/logs/log_processor.rs#L185-L188

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [reply opentelemetry-sdk/src/logs/log_processor.rs]: Applied

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [reply opentelemetry-sdk/src/logs/log_processor.rs]: I think the only other reason is the queue is closed; i've updated the messages to reflect that

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:158]: nit: remove the comment about emit as metric that was removed.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:163]: odd that we have to store this here just for logging purposes, but not an issue!

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:211]:
```suggestion
                message = "Logs were dropped due to a queue being full or other error. The count represents the total count of lost dropped in the lifetime of this BatchLogProcessor. Consider increasing the queue size and/or decrease delay between intervals."
```

Given this is for end-users, I think we cannot use "closed queue", as that is strictly internal implementation that we use a channel which has the possibility of being closed. See if my above suggestion is good. Happy to get other feedback as well.
Also, we can improve the messaging in the future also based on feedbacks we get, so no need to block this PR.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:183]:
```suggestion
            // Increment dropped logs count. The first time we have to drop a log,
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [inline opentelemetry-sdk/src/logs/log_processor.rs:159]:
```suggestion
```

**Review by cijothomas — APPROVED** [review]
Thanks! 
Left some suggestions for improved log messages.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [reply opentelemetry-sdk/src/logs/log_processor.rs:211]: LGTM. I defer entirely to your experience on the way we speak to users!

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (scottgerring) [reply opentelemetry-sdk/src/logs/log_processor.rs:158]:
```suggestion
// Track dropped logs - we'll log this at shutdown
```

**Comment by scottgerring** [comment]
Changes applied - should be good to go once the builds are done !

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (utpilla) [inline opentelemetry-sdk/src/logs/log_processor.rs:210]:
```suggestion
                message = "Logs were dropped due to a queue being full or other error. The count represents the total count of log records dropped in the lifetime of this BatchLogProcessor. Consider increasing the queue size and/or decrease delay between intervals."
```

**Comment by cijothomas** [comment]
@lalitb @utpilla Could you take a look? If I get one more approval, I plan to include this for today's release (0.27.1)

File: opentelemetry-sdk/src/trace/span_processor.rs
Comment (utpilla) [inline opentelemetry-sdk/src/trace/span_processor.rs:265]: Let's be consistent with the message in `BatchLogProcessor`.
```suggestion
                    message = "Beginning to drop span messages due to full/internal errors. No further log will be emitted for further drops until Shutdown. During Shutdown time, a log will be emitted with exact count of total spans dropped.");
```

**Review by utpilla — APPROVED** [review]
Left a suggestion to be consistent with the message content for both the batch processor.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (lalitb) [reply opentelemetry-sdk/src/logs/log_processor.rs:162]: I believe we can avoid this by moving the logging of dropped logs (otel_warn!) from the shutdown() method to the worker's Shutdown message processing.

**Comment by lalitb** [comment]
Would be good to include a unit test to validate that the dropped logs count is correctly reflected during shutdown. Since the field is private, we could consider adding a test-only accessor method to retrieve it. That said, if this is time-sensitive and needs to be included in today's release, can be done separately.

**Comment by cijothomas** [comment]
> Would be good to include a unit test to validate that the dropped logs count is correctly reflected during shutdown. Since the field is private, we could consider adding a test-only accessor method to retrieve it. That said, if this is time-sensitive and needs to be included in today's release, can be done separately.

I think we have orchestrated such a test in OTel .NET, so we can copy some ideas.

**Comment by cijothomas** [comment]
@scottgerring I edited the PR Title ("chore: Track dropped spans and logs due to full buffer"), to remove "chore" from it.

**Pull request merged**
